compaction: reshape sstables within compaction groups

lkshminarayanan opened the pull request:
For tables using tablet based replication strategies, the sstables should be reshaped only within the compaction groups they belong to. The shard_reshaping_compaction_task_impl now groups the sstables based on their compaction groups before reshaping them.

lkshminarayanan commented:
Hi @raphaelsc, need your comments on this. I have updated the existing logic to take the compaction groups into account while reshaping them. This will work for non tablet tables as well. Is this the right approach?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6773/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6773/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8000/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6773/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6789/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6789/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8014/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6789/)
#### Failed Tests (2/25866):
- [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6789/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
- [auth_cluster.test_password_login_message.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6789/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.1)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider6.cloudius-systems.com

raphaelsc commented on compaction/task_manager_module.hh:
```diff
@@ -623,7 +623,9 @@ public:
-    {}
+    {
+        _total_shard_size = 0;
+    }
```
we should initialize it in table_reshaping_compaction_task_impl::run() instead, and have shard_reshaping_compaction_task_impl::run() only add to it

lkshminarayanan replied:
Done.

raphaelsc commented on compaction/task_manager_module.cc:
```diff
@@ -596,15 +596,25 @@ future<> table_reshaping_compaction_task_impl::run() {
+        auto compaction_group_id = table.get_compaction_group_id_for_sstable(sstable);
+        sstables_grouped_by_compaction_group[compaction_group_id].insert(sstable);
+    }
 
```
I think we can simplify it if the original function `shard_reshaping_compaction_task_impl::run()` becomes:

`shard_reshaping_compaction_task_impl::reshape_compaction_group(std::unordered_set<sstables::shared_sstable> ssts);`

And the original `shard_reshaping_compaction_task_impl::run()`  is simply mapping sstables in _dir.get_unshared_local_sstables() to groups, and calling reshape_compaction_group(...) on each.

Also, there's no need to reshape in parallel, as compaction manager will serialize each reshape anyway. So we'd only be queueing work that won't be dispatched anyway.

raphaelsc commented on replica/distributed_loader.cc:
```diff
@@ -391,14 +391,10 @@ future<> table_populator::populate_subdir(sstables::sstable_state state, allow_o
+        auto gen = _global_table->calculate_generation_for_new_table();
+        return make_sstable(*_global_table, state, gen, _highest_version);
+    }, eligible_for_reshape_on_boot);
 
```
can you please write a python test for it (in core itself, not a dtest)? 

1) single cpu node with tablet enabled table containing 2-4 tablets
2) disable auto compaction
3) generate more than max_threshold (defaults to 32) sstables with help of 'nodetool flush', each sstable having at least a few keys to span multiple tablets
4) restart node, wait for reshape completed message in the log (there's a log watcher function).

lkshminarayanan replied:
Done. Added a new testcase to topology_custom/test_tablet.py.

denesb commented:
@lkshminarayanan when pushing a new version, please produce a changelog, describing what changed compared to the previous version. This helps reviewer(s) focus on just the changed areas, speeding up the review process.

lkshminarayanan commented:
@denesb sure!

Here is the changelog :
 - Changed the fix based on this [comment](https://github.com/scylladb/scylladb/pull/17395#discussion_r1494916136)
   - renamed `shard_reshaping_compaction_task_impl::run()` to `reshape_compaction_group()` retaining most of the existing behaviour but it now runs on the given list of sstables
   - added a new `shard_reshaping_compaction_task_impl::run()` method that now splits the sstables into compaction groups and calls `reshape_compaction_group()` on them one by one.
 - Added a new testcase in `topology_custom/test_tablets.py` and added a few supporting functions for that in `test/pylib/restclient`
 - `_total_shard_size` is now initialised inside `table_reshaping_compaction_task_impl::run()` and only added to inside `shard_reshaping_compaction_task_impl::run()`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6953/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6953/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/103/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8140/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6953/)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented on compaction/task_manager_module.cc:
```diff
@@ -621,8 +635,14 @@ future<> shard_reshaping_compaction_task_impl::run() {
                 sstables::compaction_result result = co_await sstables::compact_sstables(std::move(desc), info, table.as_table_state(), progress_monitor);
+                // update the sstables_in_cg set with new sstables and remove the reshaped ones
+                for (auto &sst : sstlist) {
+                    sstables_in_cg.erase(sst);
```
nit: & close to auto instead.

lkshminarayanan replied:
done.

raphaelsc commented on test/topology_custom/test_tablets.py:
```diff
@@ -61,3 +61,37 @@ async def test_tablet_cannot_decommision_below_replication_factor(manager: Manag
+
+    await log.wait_for("Reshape test.test .* Reshaped 32 sstables to .*", mark, 30)
+    sstable_info = await manager.api.get_sstable_info(server.ip_addr, "test", "test")
+    assert len(sstable_info[0]['sstables']) == number_of_tablets
```
great

raphaelsc reviewed (approved):
looks good

lkshminarayanan commented:
V3 of the PR fixes [this](https://github.com/scylladb/scylladb/pull/17395#discussion_r1500588408) issue.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6959/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6959/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/106/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8143/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6959/)
#### Build Details:
- Duration: 2 hr 12 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6969/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6969/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8149/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6969/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/112/)
#### Build Details:
- Duration: 2 hr 2 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint we should be good to merge

avikivity commented:
> @scylladb/scylla-maint we should be good to merge

No

bhalevy commented:
> > @scylladb/scylla-maint we should be good to merge
> 
> No

Due to the merge freeze, or is there something missing?

bhalevy commented:
> > > > @scylladb/scylla-maint we should be good to merge
> > > 
> > > 
> > > No
> > 
> > 
> > Due to the merge freeze, or is there something missing?
> 
> Isn't the merge freeze enough? No point in pinging something that can't be merged.

Yeah, I was referring to the readiness of this PR, in isolation.
But you're right.  I didn't take into the account the merge freeze.
